Provide compact numbers for y-axis tick values

JesseSDevaney opened the pull request:
- Closes: https://github.com/metabase/metabase/issues/41981

**Demo**

https://github.com/metabase/metabase/assets/22608765/184c0d4d-efdc-4a14-aee0-ce090246a656

- [ ] Tests/Visual Variants should be added to validate the new changes
- [ ] All other features should continue to work

JesseSDevaney committed:
refactor function

JesseSDevaney committed:
remove "graph.show_values" conditional for "graph.label_value_formatting" section

- removed this conditional since this formatting option will now apply to the y-axis as well

JesseSDevaney committed:
always show auto formatting

- since it now affects y-axis formatting as well, always show this setting

JesseSDevaney committed:
sync axis compact formatting with series/stacked/waterfall compact formatting

JesseSDevaney committed:
compact y-axis tick values even if no values are shown on data points

JesseSDevaney committed:
adjust auto compacting length difference

JesseSDevaney committed:
add stories for compacting y-axis

Commit:
Update Loki Snapshots

JesseSDevaney committed:
update bar-auto-formatting-full

alxnddr commented on frontend/src/metabase/visualizations/echarts/cartesian/model/axis.ts:
```diff
@@ -525,6 +530,7 @@ export function getYAxesModels(
   columnByDataKey: Record<DataKey, DatasetColumn>,
   isAutoSplitSupported: boolean,
   stackModels: StackModel[],
+  compactSeriesByDataKey: DataKey[],
```
Nit: I'd call this `compactSeriesDataKeys` because "a by y" is frequently used for dictionaries/maps

alxnddr commented on frontend/src/metabase/visualizations/echarts/cartesian/model/series.ts:
```diff
@@ -324,7 +324,7 @@ function shouldRenderCompact(
   };
 
-  return getAvgLength(true) + 3 < getAvgLength(false);
+  return getAvgLength(true) + 2 < getAvgLength(false);
```
Why do we want to change this threshold?

JesseSDevaney replied:
I was playing around with it and accidentally committed it. I put in a commit to revert this change.

alxnddr commented on frontend/src/metabase/visualizations/echarts/cartesian/model/series.ts:
```diff
@@ -333,15 +333,19 @@ export const getStackedLabelsFormatters = (
+  compactStackedSeriesByDataKey: DataKey[];
+} => {
   const formatters: StackedSeriesFormatters = {};
+  const compactStackedSeriesByDataKey: DataKey[] = [];
```
One more place with dictionary like naming.
Also, should it be stack name "area" | "bar" here instead of the first series?

JesseSDevaney replied:
It works either way. 

I can perform the "should we compact the y-axis" check by:
1. which axis the `stackModel` belongs to
2. by taking a `dataKey` that is a part of that stack (e.g. the first data key in that stacked series) and checking which axis it belongs to.

Since `getSeriesLabelsFormatters` is going to the use the 2nd method of checking by `dataKey` (no other option is available), I opted to use `dataKey`s with `getStackedLabelsFormatters`, so that I do not have to keep track of two different paradigms. I can merge them into one array since they are the same concept and let that one check to do all the work.

---
I do not believe there are any performance concerns, so I plan to keep it that way unless you see a different concern.

alxnddr reviewed (commented):
Looks good! I left a couple if comments, please have a look

JesseSDevaney committed:
rename parameter

JesseSDevaney committed:
revert compact formatting threshold

JesseSDevaney committed:
rename variable

cdeweyx reviewed (approved):
Looking good, aside from the padding thing between label and axis values raised in the Slack thread.

JesseSDevaney committed:
fix E2E spec y-axis tick value check

- since tick values can now be compacted, this test had to be updated because it fit within those conditions for compact formatting

JesseSDevaney committed:
refactor y-axis compact formatting check

JesseSDevaney committed:
improve padding between y-axis label and tick-values

JesseSDevaney commented:
@cdeweyx 

### Y-Axis Label Padding
**Before**

![Screenshot from 2024-05-15 18-03-54](https://github.com/metabase/metabase/assets/22608765/7e5e0da0-acd9-4bb9-a3d5-7ff13e39397f)


**After**

![Screenshot from 2024-05-15 18-03-44](https://github.com/metabase/metabase/assets/22608765/18be3b53-2953-4218-990f-95a308931735)

JesseSDevaney committed:
update loki visualization files

- since the margin between the y-axis label and y-axis tick values was increased, all of the loki files are being updated

alxnddr commented:
@JesseSDevaney thanks for addressing my feedback. I have one new comment about padding, we should not increase it generally and instead make a targeted solution for such cases. 
The padding was good, but we did not have compact formatting before which let us assume that widest y-axis ticks are the min/max values of the extent and we used them for the layout calculation. On the before variant you can see that if all ticks were less than "1.8M" width in length it would be fine.
<img width="117" alt="Screenshot 2024-05-15 at 7 35 47 PM" src="https://github.com/metabase/metabase/assets/14301985/e330e20e-54c7-4e59-8318-8d3c07060136">

So I believe not we need to handle this particular case when y-axis ticks are compact and instead of measuring it like Math.max(minValueWidth, maxValueWidth), we need to take a few values in between like 5 of them and measue Math.max(...assumedValuesWidths). What do you think?

JesseSDevaney committed:
revert axisNameMargin change

JesseSDevaney committed:
fix padding

JesseSDevaney committed:
update loki snapshots

JesseSDevaney commented:
> @JesseSDevaney thanks for addressing my feedback. I have one new comment about padding, we should not increase it generally and instead make a targeted solution for such cases. The padding was good, but we did not have compact formatting before which let us assume that widest y-axis ticks are the min/max values of the extent and we used them for the layout calculation. On the before variant you can see that if all ticks were less than "1.8M" width in length it would be fine. <img alt="Screenshot 2024-05-15 at 7 35 47 PM" width="117" src="https://private-user-images.githubusercontent.com/14301985/330978237-e330e20e-54c7-4e59-8318-8d3c07060136.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU4MTkxODcsIm5iZiI6MTcxNTgxODg4NywicGF0aCI6Ii8xNDMwMTk4NS8zMzA5NzgyMzctZTMzMGUyMGUtNTRjNy00ZTU5LTgzMTgtOGQzYzA3MDYwMTM2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE2VDAwMjEyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNmYmUzMDgxN2ZiMTMzZmYyZDI4ZWZmNTYyOTI0YWIxYjJhZTEyNjEyN2JlYmQzOWQ5ODQyYmU2NGVlMDFiNTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Z6-qMKyg7yAGlvB4ja2jVMhCteli41gEZKIn-EONWW0">
> 
> So I believe not we need to handle this particular case when y-axis ticks are compact and instead of measuring it like Math.max(minValueWidth, maxValueWidth), we need to take a few values in between like 5 of them and measue Math.max(...assumedValuesWidths). What do you think?

@alxnddr Thank you for this! :heart: 

I did know that we were calculating the padding ourselves based on the tick text width.

Please check out the new changes and let me know what you think.

---
**New Before**

![Screenshot from 2024-05-15 20-22-16](https://github.com/metabase/metabase/assets/22608765/95ba416b-6bb3-42b5-bbec-42abd3b16014)


**New After**

![Screenshot from 2024-05-15 20-22-26](https://github.com/metabase/metabase/assets/22608765/9e3e0ca4-2c17-4e3e-8934-e4488d40a4ed)

alxnddr commented on frontend/src/metabase/visualizations/echarts/cartesian/chart-measurements/index.ts:
```diff
@@ -25,6 +25,24 @@ import type {
+
+  const numCandidatesToAdd = 4;
+  const range = max - min;
+  Array.from({ length: numCandidatesToAdd }).forEach((_, i) => {
```
`Array.from({ length: ... })` is a bit non-idiomatic way. I think overall this function can be simplified:

```
if (min === max) {
  return [min, max]
}

const stepsCount = 4
const step = (max - min) / stepsCount
const middleValues = [];

for (let i = 0; i < stepCount; i++) {
  middleValues.push(min + step * i)
}

return [...middleValues, ...range]
```

JesseSDevaney replied:
Thank you for this! I altered it slightly, so that it is more correct.

alxnddr commented on frontend/src/metabase/visualizations/echarts/cartesian/chart-measurements/index.ts:
```diff
@@ -40,9 +58,10 @@ const getYAxisTicksWidth = (
+  const [min, max] = axisModel.extent.map(extent =>
     yAxisScaleTransforms.fromEChartsAxisValue(extent),
   );
+  const valuesToMeasure = getValuesToMeasure(min, max);
```
Let's apply it only when the compact formatting is enabled

JesseSDevaney committed:
Merge branch 'move-data-label-formatting-to-model' into feature-41981/provide-compact-numbers-for-y-axis-tick-values

JesseSDevaney committed:
refactor getValuesToMeasure

JesseSDevaney committed:
conditionally test extra y-tick values

alxnddr reviewed (approved):
Looks great 👍
